include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -2055,6 +2055,202 @@ __parallel_reduce_by_segment(oneapi::dpl::__internal::__device_backend_tag, _Exe
+                __mask_view_shifted, __temp_view_shifted)
+                .wait();
+        }
+        using _ScanInitType =
+            oneapi::dpl::__internal::__value_t<decltype(oneapi::dpl::__ranges::zip_view(__temp_view, __mask_view))>;
+        __parallel_transform_scan(
+            oneapi::dpl::__internal::__device_backend_tag{}, std::forward<_ExecutionPolicy>(__exec),
+            oneapi::dpl::__ranges::zip_view(__temp_view, __mask_view),
+            oneapi::dpl::__ranges::zip_view(std::forward<_Range3>(__out_values), __mask_view), __n,
+            oneapi::dpl::identity{},
+            oneapi::dpl::unseq_backend::__init_value<_ScanInitType>{
+                oneapi::dpl::__internal::make_tuple(__init.__value, _FlagType(1))},
+            oneapi::dpl::__internal::__segmented_scan_fun<_BinaryOperator, _FlagType, _BinaryOperator>{__binary_op},
+            /*_Inclusive*/ std::true_type{})
+            .wait();
+    }
+}
+
+template <bool __is_inclusive, typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3,
+          typename _BinaryPredicate, typename _BinaryOperator, typename _InitType>
+void
+__parallel_scan_by_segment(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&& __exec, _Range1&& __keys,
+                           _Range2&& __values, _Range3&& __out_values, _BinaryPredicate __binary_pred,
+                           _BinaryOperator __binary_op, _InitType __init)
+{
+    using _CustomName = oneapi::dpl::__internal::__policy_kernel_name<_ExecutionPolicy>;
+    using _ValueType = oneapi::dpl::__internal::__value_t<_Range2>;
+    assert(__keys.size() > 0);
+
+    if constexpr (std::is_trivially_copyable_v<_ValueType>)
+    {
+        sycl::queue __q_local = __exec.queue();
+        if (oneapi::dpl::__par_backend_hetero::__is_gpu_with_reduce_then_scan_sg_sz(__q_local))
+        {
+            __parallel_scan_by_segment_reduce_then_scan<_CustomName, __is_inclusive>(
+                __q_local, std::forward<_Range1>(__keys), std::forward<_Range2>(__values),
+                std::forward<_Range3>(__out_values), __binary_pred, __binary_op, __init)
+                .wait();
+            return;
+        }
+    }
+    // Implicit synchronization in this call. We need to wrap the policy as the implementation may still call
+    // reduce-then-scan and needs to avoid duplicate kernel names.
+    __parallel_scan_by_segment_fallback<_CustomName, __is_inclusive>(
+        oneapi::dpl::__internal::__device_backend_tag{},
+        oneapi::dpl::__par_backend_hetero::make_wrapped_policy<__scan_by_seg_fallback>(
+            std::forward<_ExecutionPolicy>(__exec)),
+        std::forward<_Range1>(__keys), std::forward<_Range2>(__values), std::forward<_Range3>(__out_values),
+        __binary_pred, __binary_op, __init,
+        oneapi::dpl::unseq_backend::__has_known_identity<_BinaryOperator, _ValueType>{});
```

Comment:
I was thinking about if it is worth it to keep the path where we have a "known identity", but don't qualify for `reduce_then_scan`.

I think that all types with known identities with some binary operator are trivially copyable (arithmetic types and bool).  Therefore, this is effectively a sycl CPU-specific algorithm (or for GPUs which don't support subgroup size of 32).  I think this is still worthwhile, and likely is a large improvement over the no known identity fallback for those cases.

Nothing to do act on here, just wanted to post my thoughts...